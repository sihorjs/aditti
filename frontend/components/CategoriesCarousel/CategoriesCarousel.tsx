import React, { Component, RefObject } from "react";
import Swiper from "swiper";
import "swiper/dist/css/swiper.min.css";
import "./CategoriesCarousel.less";
import { CategoryType } from "types/categories";

type Props = {
    categories: Array<CategoryType>
}

class CategoriesCarousel extends Component<Props> {
    private categories: RefObject<HTMLDivElement>;
    private slidesPerView: number;

    private constructor(props: Props) {
        super(props);

        this.categories = React.createRef();
        this.slidesPerView = 3;
    }

    componentDidMount() {
        new Swiper(this.categories.current, {
            slidesPerView: this.slidesPerView
        });
    }

    render() {
        return this.props.categories.length ? (
            <div className="swiper-container" ref={this.categories}>
                <div className="swiper-wrapper">
                    {this.props.categories.map(category => {
                        return (
                            <div
                                className="swiper-slide category"
                                key={category.name}
                            >
                                <div className="category__img-container">
                                    <img
                                        alt={category.name}
                                        className="category__img"
                                        src={category.src}
                                    />
                                </div>
                                <div className="category__info">
                                    <div className="category__name">
                                        {category.name}
                                    </div>
                                    <button
                                        className="category__btn"
                                        name="shop"
                                        type="button"
                                    >
                                        Buy
                                    </button>
                                </div>
                            </div>
                        );
                    })}
                </div>
            </div>
        ) : null;
    }
}

export default CategoriesCarousel;
