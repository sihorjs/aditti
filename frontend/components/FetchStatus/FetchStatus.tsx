import React from "react";
import LoadingMessage from "components/LoadingMessage";
import ErrorMessage from "components/ErrorMessage";
import { FetchStatusType } from "frontend/types";

function FetchStatus(props: FetchStatusType) {
    if (props.isLoading) {
        return <LoadingMessage />;
    } else if (props.hasError) {
        return <ErrorMessage />;
    }

    return null;
}

export default FetchStatus;
