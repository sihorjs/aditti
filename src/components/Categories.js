import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import Swiper from "swiper";
import "swiper/dist/css/swiper.min.css";

class Categories extends Component {
    constructor(props) {
        super(props);

        this.categories = React.createRef();
    }

    componentDidMount() {
        new Swiper(this.categories.current, { slidesPerView: 3 });
    }

    render() {
        return (
            <div className="categories swiper-container" ref={this.categories}>
                <div className="swiper-wrapper">
                    {this.props.categories.map(category => {
                        return (
                            <div
                                className="swiper-slide category"
                                key={category.name}
                            >
                                <div className="category-img">
                                    <img
                                        src={category.src}
                                        alt={category.name}
                                    />
                                </div>
                                <div className="category-title">
                                    <span>{category.name}</span>
                                    <button name="shop">Buy</button>
                                </div>
                            </div>
                        );
                    })}
                </div>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        categories: state.catalogue.data.categories
    };
}

Categories.propTypes = {
    categories: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            src: PropTypes.string.isRequired
        })
    ).isRequired
};

export default connect(mapStateToProps)(Categories);
