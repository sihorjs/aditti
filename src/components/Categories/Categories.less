@import url("less/index.less");

.category {
    display: flex;
    width: 350px;
    height: 170px;

    &:not(:last-child) {
        border-right: solid 1px @gray;
    }

    &__img-container {
        display: flex;
        justify-content: center;
        align-items: center;
        width: 50%;
        box-sizing: border-box;
        padding: 5px;
    }

    &__img {
        max-width: 100%;
        height: auto;
    }

    &__info {
        width: 50%;
        padding-top: 50px;
        font-family: @heading-font;
    }

    &__name {
        margin-bottom: 5px;
        font-size: 27px;
        color: @gray;
    }

    &__btn {
        width: 60px;
        height: 35px;
        border: none;
        font: 19px @heading-font;
        color: @white;
        background-color: @cyan;
        border-radius: 5px;
        cursor: pointer;

        &:hover {
            background-color: @darkcyan;
        }
    }
}
