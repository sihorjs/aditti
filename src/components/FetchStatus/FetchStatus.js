import React from "react";
import PropTypes from "prop-types";

import LoadingMessage from "components/LoadingMessage";
import ErrorMessage from "components/ErrorMessage";

function FetchStatus(props) {
    if (props.isLoading) {
        return <LoadingMessage />;
    } else if (props.hasError) {
        return <ErrorMessage />;
    }

    return null;
}

FetchStatus.propTypes = {
    hasError: PropTypes.bool.isRequired,
    isLoading: PropTypes.bool.isRequired
};

export default FetchStatus;
