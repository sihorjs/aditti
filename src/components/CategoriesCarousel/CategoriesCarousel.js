import React, { Component } from "react";
import PropTypes from "prop-types";
import Swiper from "swiper";
import "swiper/dist/css/swiper.min.css";
import "./CategoriesCarousel.less";

class CategoriesCarousel extends Component {
    constructor(props) {
        super(props);

        this.categories = React.createRef();
        this.slidesPerView = 3;
    }

    componentDidMount() {
        new Swiper(this.categories.current, {
            slidesPerView: this.slidesPerView
        });
    }

    render() {
        return this.props.categories.length ? (
            <div className="swiper-container" ref={this.categories}>
                <div className="swiper-wrapper">
                    {this.props.categories.map(category => {
                        return (
                            <div
                                className="swiper-slide category"
                                key={category.name}
                            >
                                <div className="category__img-container">
                                    <img
                                        alt={category.name}
                                        className="category__img"
                                        src={category.src}
                                    />
                                </div>
                                <div className="category__info">
                                    <div className="category__name">
                                        {category.name}
                                    </div>
                                    <button
                                        className="category__btn"
                                        name="shop"
                                        type="button"
                                    >
                                        Buy
                                    </button>
                                </div>
                            </div>
                        );
                    })}
                </div>
            </div>
        ) : null;
    }
}

CategoriesCarousel.propTypes = {
    categories: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            src: PropTypes.string.isRequired
        })
    ).isRequired
};

export default CategoriesCarousel;
