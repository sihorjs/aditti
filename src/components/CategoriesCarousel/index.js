import CategoriesCarousel from "./CategoriesCarousel";

export default CategoriesCarousel;
