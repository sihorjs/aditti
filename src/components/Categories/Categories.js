import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import Swiper from "swiper";
import "swiper/dist/css/swiper.min.css";
import "./Categories.less";

class Categories extends Component {
    constructor(props) {
        super(props);

        this.categories = React.createRef();
        this.slidesPerView = 3;
    }

    componentDidMount() {
        new Swiper(this.categories.current, {
            slidesPerView: this.slidesPerView
        });
    }

    render() {
        return (
            <section
                className="wrapper swiper-container"
                ref={this.categories}
            >
                <div className="swiper-wrapper">
                    {this.props.categories.map(category => {
                        return (
                            <div
                                className="swiper-slide category"
                                key={category.name}
                            >
                                <div className="category__img-container">
                                    <img
                                        className="category__img"
                                        src={category.src}
                                        alt={category.name}
                                    />
                                </div>
                                <div className="category__info">
                                    <div className="category__name">
                                        {category.name}
                                    </div>
                                    <button
                                        className="category__btn"
                                        name="shop"
                                        type="button"
                                    >
                                        Buy
                                    </button>
                                </div>
                            </div>
                        );
                    })}
                </div>
            </section>
        );
    }
}

function mapStateToProps(state) {
    return {
        categories: state.catalogue.data.categories
    };
}

Categories.propTypes = {
    categories: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            src: PropTypes.string.isRequired
        })
    ).isRequired
};

export default connect(mapStateToProps)(Categories);
