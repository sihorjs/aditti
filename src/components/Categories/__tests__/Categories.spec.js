import React from "react";
import { shallow } from "enzyme";
import { Categories } from "../Categories";
import FetchStatus from "components/FetchStatus";
import CategoriesCarousel from "components/CategoriesCarousel";

describe("<Categories />", () => {
    const onGetCategories = jest.fn();

    afterEach(() => jest.clearAllMocks());

    it("renders without crash", () => {
        const tree = shallow(
            <Categories
                categories={[]}
                hasError={false}
                isLoading={false}
                onGetCategories={onGetCategories}
            />
        );

        expect(tree).toMatchSnapshot();
    });

    it("starts fetch of categories in componentDidMount", () => {
        shallow(
            <Categories
                categories={[]}
                hasError={false}
                isLoading={false}
                onGetCategories={onGetCategories}
            />
        );

        expect(onGetCategories).toHaveBeenCalledTimes(1);
    });

    it("renders FetchLoading during categories loading", () => {
        const component = shallow(
            <Categories
                categories={[]}
                hasError={false}
                isLoading
                onGetCategories={onGetCategories}
            />
        );

        expect(component.type()).toEqual(FetchStatus);
    });

    it("renders FetchLoading if fetch was unsuccessful", () => {
        const component = shallow(
            <Categories
                categories={[]}
                hasError
                isLoading={false}
                onGetCategories={onGetCategories}
            />
        );

        expect(component.type()).toEqual(FetchStatus);
    });

    it(
        "renders CategoriesCarousel with categories list if fetch was finished",
        () => {
            const categories = [];
            const component = shallow(
                <Categories
                    categories={categories}
                    hasError={false}
                    isLoading={false}
                    onGetCategories={onGetCategories}
                />
            );

            expect(component.find("section.categories").exists()).toBeTruthy();
            expect(
                component
                    .find("section.categories")
                    .contains(<CategoriesCarousel categories={categories} />)
            ).toBeTruthy();
        }
    );
});
