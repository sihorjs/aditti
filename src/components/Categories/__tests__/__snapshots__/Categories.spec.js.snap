// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Categories /> renders without crash 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Categories
    categories={Array []}
    hasError={false}
    isLoading={false}
    onGetCategories={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <CategoriesCarousel
        categories={Array []}
      />,
      "className": "categories wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "categories": Array [],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <CategoriesCarousel
          categories={Array []}
        />,
        "className": "categories wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "categories": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
